lib/rouge/lexers/xpath.rb

```diff
@@ -0,0 +1,298 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+module Rouge
+  module Lexers
+    class XPath < RegexLexer
+      title 'XPath'
+      desc 'XML Path Language (XPath) 3.1'
+      tag 'xpath'
+      filenames '*.xpath'
+
+      # Terminal literals:
+      # https://www.w3.org/TR/xpath-31/#terminal-symbols
+
+      digits = /[0-9]+/
```

Comment:
Thank you so much for the detailed feedback! It's all super instructive, so really, thank you for taking the time to type this out :smiley:

If I understand it correctly:

|          | Object               | Instance                           |
| -------- | -------------------- | ---------------------------------- |
| Method   | `def self.foo`       | `def foo`                          |
| Variable | `@@foo`              | `@foo` (instance) or `foo` (local) |

(I googled my way to `@@foo`  just to complete the table). 

I've converted all the local variable assignments in 2ec2ed28993856bf7bbda6bcc26ccae45dbd46dc

Reply:
Ahh... yes, variables and scope in Ruby. OK, so this is really confusing but Ruby has 4 different kinds of scoped variables:

1. **Class Variables**
   These have the form `@@foo`. They are shared between all instances of a class.

2. **Singleton Class Variables**
   These have the form `@foo` but they are defined in the scope of the singleton class. When you define a 'class' method (ie. `def self.foo`) in Ruby this is really a method on the singleton class. Since you are within the scope of the singleton class inside this method, a reference to `@foo` is a reference to the instance variable of the singleton class. These variables are not directly accessible from instances of the class.

   Now what, you might ask, is the singleton class? It hurts my brain to think about but you might find [this blog post](https://www.devalot.com/articles/2008/09/ruby-singleton) useful as an explanation.

3. **Instance Variables**
   These are the same as the instance variables you know and love from other object-oriented languages. Each instance of the class gets its own set of instance variables.

4. **Local Variables**
   These are variables that are defined and exist solely within a particular scope. The class, singleton and instance scopes can all have local variables. You were originally using local variables within the scope of the class.

So your table should look more like this:

|               | Class          | Singleton                                       | Instances  |
|---------|------------|---------------------------------|-----------|
| method |  --            |   `self.foo()`                                  | `foo()`      |
| variable |   `@@foo` |   `@foo` (in the scope of `self`) | `@foo`         |

Reply:
Thank you! This clears things up. The blog post mentioned "anonymous classes" as an alternative name for "singleton classes", which made things click for me.